File: petclinic-java-gradle-react-tailwind-ts/src/main/java/dev/pollito/petclinic_java_gradle_react_tailwind_ts/type/TypeServiceImpl.java
package dev.pollito.petclinic_java_gradle_react_tailwind_ts.type;

import dev.pollito.petclinic_java_gradle_react_tailwind_ts.events.BeforeDeleteType;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.CustomCollectors;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.NotFoundException;
import java.util.List;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class TypeServiceImpl implements TypeService {

    private final TypeRepository typeRepository;
    private final ApplicationEventPublisher publisher;

    @Override
    public List<Type> findAll() {
        return typeRepository.findAll(Sort.by("id"));
    }

    @Override
    public Type get(final Integer id) {
        return typeRepository.findById(id)
                .orElseThrow(NotFoundException::new);
    }

    @Override
    public Integer create(final Type type) {
        return typeRepository.save(type).getId();
    }

    @Override
    public void update(final Integer id, final Type type) {
        type.setId(id);
        typeRepository.findById(id).orElseThrow(NotFoundException::new);
        typeRepository.save(type);
    }

    @Override
    public void delete(final Integer id) {
        final Type type = typeRepository.findById(id)
                .orElseThrow(NotFoundException::new);
        publisher.publishEvent(new BeforeDeleteType(id));
        typeRepository.delete(type);
    }

    @Override
    public Map<Integer, Integer> getTypeValues() {
        return typeRepository.findAll(Sort.by("id"))
                .stream()
                .collect(CustomCollectors.toSortedMap(Type::getId, Type::getId));
    }
}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/type/TypeServiceImpl.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.type

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeleteType
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.CustomCollectors
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.NotFoundException
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service

@Service
class TypeServiceImpl(
    private val typeRepository: TypeRepository,
    private val publisher: ApplicationEventPublisher,
    private val typeMapper: TypeMapper,
) : TypeService {

    override fun findAll(): List<TypeDTO> {
        val types = typeRepository.findAll(Sort.by("id"))
        return types.map { type -> typeMapper.updateTypeDTO(type, TypeDTO()) }
    }

    override fun `get`(id: Int): TypeDTO = typeRepository.findById(id)
        .map { type -> typeMapper.updateTypeDTO(type, TypeDTO()) }
        .orElseThrow { NotFoundException() }

    override fun create(typeDTO: TypeDTO): Int {
        val type = Type()
        typeMapper.updateType(typeDTO, type)
        return typeRepository.save(type).id!!
    }

    override fun update(id: Int, typeDTO: TypeDTO) {
        val type = typeRepository.findById(id)
            .orElseThrow { NotFoundException() }
        typeMapper.updateType(typeDTO, type)
        typeRepository.save(type)
    }

    override fun delete(id: Int) {
        val type = typeRepository.findById(id)
            .orElseThrow { NotFoundException() }
        publisher.publishEvent(BeforeDeleteType(id))
        typeRepository.delete(type)
    }

    override fun getTypeValues(): Map<Int, Int> = typeRepository.findAll(Sort.by("id"))
        .stream()
        .collect(CustomCollectors.toSortedMap(Type::id, Type::id))
}
